common/konfig.py

```diff
@@ -0,0 +1,447 @@
+# SPDX-FileCopyrightText: © 2024 Christian BUHTZ <c.buhtz@posteo.jp>
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# This file is part of the program "Back In Time" which is released under GNU
+# General Public License v2 (GPLv2).
+# See file LICENSE or go to <https://www.gnu.org/licenses/#GPL>.
+from __future__ import annotations
+import configparser
+import getpass
+import contextlib
+import os
+from typing import Union, Any
+from pathlib import Path
+from io import StringIO, TextIOWrapper
+import singleton
+import logger
+
+# Workaround: Mostly relevant on TravisCI but not exclusively.
+# While unittesting and without regular invocation of BIT the GNU gettext
+# class-based API isn't setup yet.
+# The bigger problem with config.py is that it do use translatable strings.
+# Strings like this do not belong into a config file or its context.
+try:
+    _('Warning')
+except NameError:
+    _ = lambda val: val
+
+
+class Profile:
+    """Manages access to profile-specific configuration data."""
+    _DEFAULT_VALUES = {
+        'snapshots.mode': 'local',
+        'snapshots.ssh.port': 22,
+        'snapshots.ssh.cipher': 'default',
+        'snapshots.ssh.user': getpass.getuser(),
+        'snapshots.ssh.private_key_file':
+            str(Path('~') / '.ssh' / 'id_rsa'),
```

Comment:
The intention is not to get expanded. Because the `~` is used in the string written to the config file. That is how a users would do it in some cases. Otherwies I would have written `Path.home()`.

The separate strings (ohne string per path component/folder) is to my knowledge the regular way how the `pathlib` package is intended to be used. This is platform independence. The package decide itself how the folder/path separator is written in the current file system.

Reply:
Pathlib is platform independent, but you don't need to separate the path out to make it work. The following will be correctly converted to a Windows path on Windows:

``` python
Path('~/example/path')
```

You can test to see what would happen using `pathlib.PureWindowsPath`:

``` python
p = str(pathlib.PureWindowsPath('~/example/path'))
print(p)
# ~\example\path
```

As you can see, a path containing forward slashes is correctly converted to a valid Windows path automatically.

By using `Path('~/example/path')`, you will automatically get forward slashes on non-Windows, and back-slashes on Windows.

So, given the code above, the following is exactly equivalent:

``` python
'snapshots.ssh.private_key_file': str(Path('~/.ssh/id_rsa')),
```

On Linux, this will become:

```
'snapshots.ssh.private_key_file': '~/.ssh/id_rsa',
```

While on Windows it will become:

```
'snapshots.ssh.private_key_file': r'~\.ssh\id_rsa',
```

I'm not sure how this config is used, but you might even get away with keeping it as a Path object without converting it to string.

Reply:
Technically you are right. There is no difference. But I would like to have the syntax also platform independent and not only the behavior.

It is just a matter of habit and style. There is no strict technical reason for it.

I might be to pedantic about it because I am also not aware of any linter rule checking for something like this. 😆

Reply:
Happy to respect a style choice.